storage-node/src/cache/data_store_cache/memdisk/mod.rs

```diff
@@ -118,46 +137,108 @@ impl<R: DataStoreReplacer<MemDiskStoreReplacerKey, MemDiskStoreReplacerValue>> D
+                {
+                    Ok(Some(rx)) => {
+                        return Ok(Some(rx));
+                    }
                     Ok(None) => {
                         return Err(ParpulseError::Internal(
-                            "Memory replacer and memory store keys are inconsistent.".to_string(),
+                            "Memory replacer and memory store is inconsistent.".to_string(),
                         ))
                     }
                     Err(e) => return Err(e),
                 }
             }
         }
 
-        let mut disk_replacer = self.disk_replacer.write().await;
-        match disk_replacer.get(&remote_location) {
-            Some(replacer_value) => {
-                let data_store_key = replacer_value.as_value();
-                if let Some(data) = self.disk_store.read_data(data_store_key).await? {
-                    Ok(Some(data))
-                } else {
-                    unreachable!()
+        let data_store_key;
+        {
+            let mut disk_replacer = self.disk_replacer.lock().await;
+            match disk_replacer.get(&remote_location) {
+                Some(replacer_value) => {
+                    data_store_key = replacer_value.as_value().clone();
                 }
+                None => return Ok(None),
             }
-            None => Ok(None),
+        }
+        if let Some(data) = self
+            .disk_store
+            .read_data(&data_store_key, self.disk_replacer.clone())
+            .await?
+        {
+            Ok(Some(data))
+        } else {
+            return Err(ParpulseError::Internal(
+                "Disk replacer and disk store is inconsistent.".to_string(),
+            ));
         }
     }
 
     async fn put_data_to_cache(
-        &mut self,
+        &self,
         remote_location: String,
         _data_size: Option<usize>,
         mut data_stream: StorageReaderStream,
     ) -> ParpulseResult<usize> {
+        {
+            // If in_progress of remote_location thread fails, it will clean the data from this hash map.
+            let mut existed = false;
+            loop {
+                let status_of_keys = self.status_of_keys.read().await;
+                if let Some(((status, size), notify_ref)) = status_of_keys.get(&remote_location) {
+                    let notify;
+                    existed = true;
+                    match status {
+                        Status::Incompleted => {
+                            notify = notify_ref.clone();
+                            *notify_ref.1.lock().await += 1;
+                            // The Notified future is guaranteed to receive wakeups from notify_waiters()
+                            // as soon as it has been created, even if it has not yet been polled.
+                            let notified = notify.0.notified();
+                            drop(status_of_keys);
+                            notified.await;
```

Comment:
Maybe it's better to have a customized `CondVar` ([sth. like this](https://github.com/kaimast/tokio-condvar/blob/main/src/lib.rs)). But we can keep it as is and refactor it later.

Reply:
we can later add a github issue to record all the refactor actions?